### PR TITLE
stm32 testing tweaks

### PR DESCRIPTION
**Pull request by drowe67**
@donmr - this is an attempt to undo the multiple-merge mess I made in https://github.com/drowe67/codec2/pull/45

Aiming to get just the stm32 changes and not the sync stuff in this PR.

### CODE REVIEW

**Comment by drowe67** [comment]
tst_api_demod_700D_plain_codec is close, but some divergence in the outputs speech.  They both sound fine.  Notes on possible causes in tst_api_demod_check.  Here is a plot of the two outputs and difference.  

Seems to diverge over time, and particularly bad in unvoiced speech.

Screenshot from 2019-06-22 14-02-29

**Comment by drowe67** [comment]
@db4ple I'm chasing down a tricky bug in the fdmdv modem used for FreedDV1600 on the stm32.  After a few frames it's getting NANs in a bunch of vectors.  Have traced it to the ARM trig functions:
```
w: -0.000000 COSF(w) : 1.000000 SINF(w) 6.283029 cosf(w) 1.000000 sinf(w) -0.000000
```

**Comment by srsampson** [comment]
From the docs:

" [0 +0.9999] mapping to [0 2*pi).  The fixed-point range is chosen so that
a  value of 2*pi wraps around to 0."

https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/DSP/Source/FastMathFunctions/arm_sin_f32.c#L44

**Comment by srsampson** [comment]
Looks like CMSIS 5 fixed the bug in the library:

  /* Special case for small negative inputs */
  if ((x < 0.0f) && (x >= -1.9e-7f)) {
     return x;
  }

**Comment by drowe67** [comment]
> Looks like CMSIS 5 fixed the bug in the library: /* Special case for small negative inputs */ if ((x < 0.0f) && (x >= -1.9e-7f)) { return x; }
> […](#)
> On Sat, Jun 22, 2019 at 7:52 PM Steve ***@***.***> wrote: From the docs: " [0 +0.9999] mapping to [0 2*pi). The fixed-point range is chosen so that a value of 2*pi wraps around to 0." https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/DSP/Source/FastMathFunctions/arm_sin_f32.c#L44

Thanks @srsampson.  Prior to reading your comment I pulled it out.  We can consider putting it back (and testing **very** carefully) if there is a compelling reason (like large speed improvement).  It was only really used in the fdmdv modem, and I'm a bit wary of those functions now after a day of chasing down the bug...

File: src/debug_alloc.h [hunk 1]
```diff
@@ -20,10 +20,10 @@ register char * sp asm ("sp");
 
   static inline void * DEBUG_MALLOC(const char *func, size_t size) {
     void *ptr = malloc(size);
-    fprintf(stderr, "MALLOC: %s %p %zd", func, ptr, size);
+    fprintf(stderr, "MALLOC: %s %p %d", func, ptr, size);
 #ifdef CORTEX_M4
 
-    fprintf(stderr, " : sp %p  heap_end %p", sp, __heap_end);
+    fprintf(stderr, " : sp %p ", sp);
 #endif
```
Comment (drowe67) [inline src/debug_alloc.h]: I couldn't find the ```heap_end``` symbol ....

File: stm32/unittest/src/tst_api_demod.c [hunk 2]
```diff
@@ -100,10 +102,8 @@ int main(int argc, char *argv[]) {
     struct my_callback_state my_cb_state;
     int            frame;
     int            nread, nin, nout;
-    struct MODEM_STATS stats;
     int            sync;
```
Comment (drowe67) [inline stm32/unittest/src/tst_api_demod.c]: This has grown with my recent FreeDV 2020 adventures.  Need to do something about making this usable in embedded systems.

**Comment by hogthrob** [comment]
> 
> 
> > Looks like CMSIS 5 fixed the bug in the library: /* Special case for small negative inputs _/ if ((x < 0.0f) && (x >= -1.9e-7f)) { return x; }
> > […](#)
> > On Sat, Jun 22, 2019 at 7:52 PM Steve _**@**_.___> wrote: From the docs: " [0 +0.9999] mapping to [0 2_pi). The fixed-point range is chosen so that a value of 2_pi wraps around to 0." https://github.com/ARM-software/CMSIS_5/blob/develop/CMSIS/DSP/Source/FastMathFunctions/arm_sin_f32.c#L44
> 
> Thanks @srsampson. Prior to reading your comment I pulled it out. We can consider putting it back (and testing **very** carefully) if there is a compelling reason (like large speed improvement). It was only really used in the fdmdv modem, and I'm a bit wary of those functions now after a day of chasing down the bug...

Sorry guys, on the UHSDR we're using a different CMSIS DSP lib (newer) so this issue did not appear on our end.

**Comment by donmr** [comment]
In debug_alloc.h the format for "size" should be "%ld".

**Comment by donmr** [comment]
Here is sbrk_dbg.c which I made from the library source.  It defines __heap_limit as external instead of static so others can see it.

#include <_ansi.h>
#include <sys/types.h>
#include <stdio.h>
#include <errno.h>

/* Register name faking - works in collusion with the linker.  */
register char * stack_ptr asm ("sp");

/* Heap limit returned from SYS_HEAPINFO Angel semihost call.  */
extern uint __heap_limit;
char * __heap_end;

caddr_t 
_sbrk (int incr)
{
  extern char end asm ("end"); /* Defined by the linker.  */
  char * prev_heap_end;

  if (__heap_end == NULL)
    __heap_end = & end;
  
  prev_heap_end = __heap_end;
  
  if ((__heap_end + incr > stack_ptr)
      /* Honour heap limit if it's valid.  */
      || (__heap_limit != 0xcafedead && (uint)(__heap_end + incr) > __heap_limit))
    {
      /* Some of the libstdc++-v3 tests rely upon detecting
	 out of memory errors, so do not abort here.  */
#if 0
      extern void abort (void);

      _write (1, "_sbrk: Heap and stack collision\n", 32);
      
      abort ();
#else
      errno = ENOMEM;
      return (caddr_t) -1;
#endif
    }
 
  __heap_end += incr;
  fprintf(stderr, "new heap_end %p\n", __heap_end);

  return (caddr_t) prev_heap_end;
}

**Comment by donmr** [comment]
The changes to not require a compare are allowing a test to pass when it did not even run.  I don't yet know why it isn't running though:

$ run_stm32_tst tst_codec2_enc 1300
test full name: tst_codec2_enc_1300
----------------------- STARTING gdb/OpenOCD ----------------------------
Open On-Chip Debugger 0.10.0+dev-00796-ga4ac5615 (2019-04-14-10:41)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
debug_level: 0
adapter speed: 2000 kHz
adapter_nsrst_delay: 100
none separate
srst_only separate srst_nogate srst_open_drain connect_deassert_srst


gdb_cmds:4: Error in sourced command file:
:3333: Connection timed out.

Compare output binary data
Error opening file2 "stm_out.raw": No such file or directory
Passed

Test PASSED

Test tst_codec2_enc_1300 PASSED

**Comment by drowe67** [comment]
Thanks so how  do we replace the library version of this function with our new version?  Do we need to rebuild the standard library?

**Comment by drowe67** [comment]
Re ```$ run_stm32_tst tst_codec2_enc 1300``` - thanks, I think it's because I am not resetting gdb_cmd.  If I force the error I get the same behaivor,.  It should (i)be dying after the run test fails, and (ii) trapping the fail of the missing file in _check.  Will get on to fixing those issues ....

**Comment by drowe67** [comment]
@donmr I've had a look around but can't work out why you're getting the:
```
gdb_cmds:4: Error in sourced command file:
:3333: Connection timed out.
```
Seems to suggest it can't connect to openocd, but that appears to have started.  Working fine here. Can you take a look into this on yr machine pls?

I'll move on to  the PASS issue.

**Comment by donmr** [comment]
Just include it as a source file.  Than the library function won't get loaded.

**Comment by donmr** [comment]
Resets and a reboot fixed my problem getting gdb/openocd to run.  Something had apparently locked up the USB / udev stuff.  But it was informative to see that this failure resulted in a PASSED report.  We don't test for that sort of thing very often.

**Comment by donmr** [comment]
Having changed run_stm32_tst to default to load, with a --noload option, we need to change the run_all_*_tests scripts to remove --load and add --noload.  

I can easily do this but it the process of making more branches, PRs, and merges might be more complicated than David just making the changes in his dr-stm32-wip3 branch.

**Comment by drowe67** [comment]
> Resets and a reboot fixed my problem getting gdb/openocd to run. Something had apparently locked up the USB / udev stuff. 

Yes I've had that a few times too.

> But it was informative to see that this failure resulted in a PASSED report. We don't test for that sort of thing very often.

Agreed!  We need to trap that sort of thing.  I've added some code to do that, and tested it by unplugging the Discovery.  However it just traps ```run_stm32_prog``` exiting with non-zero, it doesn't test for the existence of  stm_out.raw, which was the second reason the test logic PASSED.  To address this:
1. We could add checks for stm_out.raw to every ```*_check``` script
1. Or a check for ```stm_out.raw``` in ```run_stm32_tst```, but this assumes every test will output this file.
1. Or just assume the above check that ```run_stm32_prog``` worked is enough coverage.

File: stm32/unittest/src/tst_api_demod.c [hunk 3]
```diff
@@ -93,20 +97,25 @@ void my_datatx(void *callback_state, unsigned char *packet, size_t *size) {
     *size = 0;
 }
 
+#define SPARE_RAM 10000
 
 int main(int argc, char *argv[]) {
+    char           dummy[SPARE_RAM];
     int            f_cfg, f_in, f_out;
```
Comment (drowe67) [inline stm32/unittest/src/tst_api_demod.c]: @donmr I added this dummy buffer in order to trigger an out of memory assert should the complete program have less that 10k free.  Any other ideas on how to do this?

**Comment by drowe67** [comment]
OK just did a clean build and all 23 checks passing.  Takes a while due to the slow semi-hosting but it's a great set of tests - well done @donmr :+1: 

Will look into adding the custom ```sbrk_dbg.c```

@donmr would you like to work on ```sm1000_main.c``` integration, or would you like me to do that?  I'm keen to get it running on the sm1000 in the next few weeks.  Am happy to do the work if you are busy, or happy to work on other things if you'd like to do it.

I have the current (1600 only) version building and running fine with a TTY in https://github.com/drowe67/codec2/pull/42

File: stm32/unittest/scripts/tst_api_demod_check [hunk 4]
```diff
@@ -37,24 +50,81 @@ case "${TEST_OPT}" in
...
+	compare_energy;
+
+	# make sure execution time stays within bounds
+	execution_time=mktmp
+	cat stdout.log | sed  -n "s/.*freedv_rx \([0-9..]*\) msecs/\1/p" > $execution_time
+	python -c "import sys; import numpy as np;  x=np.loadtxt(\"$execution_time\"); print(\"execution time max:: %5.2f mean: %5.2f ms\" % (np.max(x), np.mean(x))); sys.exit(1) if np.max(x) < 80.0 else sys.exit(0)"
+	if [[ $? -eq 1 ]]; then echo "execution time OK"; 
+        else echo "BAD"; 
+     	   let Fails=($Fails + 1)
+        fi
```
Comment (drowe67) [inline stm32/unittest/scripts/tst_api_demod_check]: @donmr I added code to check execution time.  Had to place a limit on LPDC decoder iterations to get it into real time.

**Comment by donmr** [comment]
I would check for an output of at least some minimum size in each
*_check script that is not doing a compare.

We could also scan the gdb log for error messages.

Don

**Comment by donmr** [comment]
I'm not sure that will get allocated unless it is used.  These compilers
are too smart.  How about a malloc near the end (but before any free's)?

**Comment by donmr** [comment]
If you have time to do sm1000 integration that is great.  I could get to
it but have been caught up in several unplanned activities, it's summer
here now.

I'll try to at least follow what is happening and do some testing.

Don

**Comment by drowe67** [comment]
Yes it does get used, when i get to 15000 bytes it falls over.  Note I've made it the first variable, and the stack grows down.  A malloc at the end is a  good idea too :+1: 

> I'm not sure that will get allocated unless it is used. These compilers are too smart. How about a malloc near the end (but before any free's)?
> […](#)
> On Fri, Jun 28, 2019 at 03:44:11PM -0700, drowe67 wrote: drowe67 commented on this pull request. > int main(int argc, char *argv[]) { + char dummy[SPARE_RAM]; int f_cfg, f_in, f_out; @donmr I added this dummy buffer in order to trigger an out of memory assert should the complete program have less that 10k free. Any other ideas on how to do this? -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#46 (review)](https://github.com/drowe67/codec2/pull/46#pullrequestreview-255999953)

**Comment by drowe67** [comment]
> I would check for an output of at least some minimum size in each *_check script that is not doing a compare. We could also scan the gdb log for error messages. Don

Yes, I'm grepping for ```assertions``` as well, no point continuing to the ```_check``` script if it's bombed with an assertion.

**Comment by drowe67** [comment]
@donmr could you pls confirm the 23 tests work for you?  I'll them merge this rather large PR and we can do future improvements in a new one.

**Comment by donmr** [comment]
Now its warning that nmemb is also a size_t, in debug_alloc.h

**Comment by donmr** [comment]
And I was wrong, size_t should be %zd not %ld.

**Comment by donmr** [comment]
The 16 tests run by "run_all_stm32_tests" pass for me.  Also tst_api_mod_700D_TEST, tst_api_mod_700D_CODEC,  tst_api_demod_700D_plain_test, 
tst_api_demod_700D_AGWN_test, tst_api_demod_700D_AWGN_test, tst_api_demod_700D_AWGN_codec, and tst_api_demod_1600_plain_codec.

**Comment by drowe67** [comment]
> The 16 tests run by "run_all_stm32_tests" pass for me. Also tst_api_mod_700D_TEST, tst_api_mod_700D_CODEC, tst_api_demod_700D_plain_test,
> tst_api_demod_700D_AGWN_test, tst_api_demod_700D_AWGN_test, tst_api_demod_700D_AWGN_codec, and tst_api_demod_1600_plain_codec.

I just use ```make all test``` in the build dir ```~/codec2/stm32/build_stm32``` to run them all.  But either way, looks like you are getting the same results :+1:

**Pull request merged**